codes/ruby/chapter_backtracking/n_queens.rb

```diff
@@ -0,0 +1,60 @@
+=begin
+File: n_queens.rb
+Created Time: 2024-04-07
+Author: hoangtien2k3 (hoangtien2k3qx1@gmail.com)
+=end
+
+module Backtracking
+  # 回溯算法：n 皇后
+  def self.backtrack(row, n, state, res, cols, diags1, diags2)
+    # 当放置完所有行时，记录解
+    if row == n
+      copy_state = state.map(&:dup)
+      res << copy_state
+      return
+    end
+
+    # 遍历所有列
+    (0...n).each do |col|
+      # 计算该格子对应的主对角线和次对角线
+      diag1 = row - col + n - 1
+      diag2 = row + col
+      # 剪枝：不允许该格子所在列、主对角线、次对角线上存在皇后
+      if !cols[col] && !diags1[diag1] && !diags2[diag2]
+        # 尝试：将皇后放置在该格子
+        state[row][col] = 'Q'
+        cols[col] = diags1[diag1] = diags2[diag2] = true
+        # 放置下一行
+        backtrack(row + 1, n, state, res, cols, diags1, diags2)
+        # 回退：将该格子恢复为空位
+        state[row][col] = '#'
+        cols[col] = diags1[diag1] = diags2[diag2] = false
+      end
+    end
+  end
+end
+
+# 求解 n 皇后
```

Comment:
Please modify all the heading comments as the following. You can check the corresponding heading comments in the Python files (the `""" """` block)

```rb
### 求解 n 皇后 ###
```